code/datums/emote.dm

```diff
@@ -0,0 +1,532 @@
+
+/datum/emote/New()
+	if(message_param && !param_desc)
+		CRASH("emote [src] was given a message parameter without a description.")
+	if(ispath(mob_type_allowed_typecache))
+		switch(mob_type_allowed_typecache)
+			if(/mob)
+				mob_type_allowed_typecache = GLOB.typecache_mob
+			if(/mob/living)
+				mob_type_allowed_typecache = GLOB.typecache_living
+			else
+				mob_type_allowed_typecache = typecacheof(mob_type_allowed_typecache)
+	else
+		mob_type_allowed_typecache = typecacheof(mob_type_allowed_typecache)
+	mob_type_blacklist_typecache = typecacheof(mob_type_blacklist_typecache)
+	mob_type_ignore_stat_typecache = typecacheof(mob_type_ignore_stat_typecache)
+	species_type_whitelist_typecache = typecacheof(species_type_whitelist_typecache)
+
+/**
+ * Handles the modifications and execution of emotes.
+ *
+ * In general, what this does:
+ * - Checks if the user can run the emote at all
+ * - Checks and applies the message parameter, if it exists
+ * - Replaces pronouns with a mob's specific pronouns
+ * - Checks for and plays sound if the emote supports it
+ * - Sends the emote to users
+ * - Runechats the emote
+ *
+ * Arguments:
+ * * user - Person that is trying to send the emote.
+ * * params - Parameters added after the emote.
+ * * type_override - Override to the current emote_type.
+ * * intentional - Bool that says whether the emote was forced (FALSE) or not (TRUE).
+ *
+ * Returns TRUE if it was able to run the emote, FALSE otherwise.
+ */
+/datum/emote/proc/run_emote(mob/user, params, type_override, intentional = FALSE)
+	. = TRUE
+	if(!can_run_emote(user, TRUE, intentional))
+		return FALSE
+	var/msg = select_message_type(user, message, intentional)
+	// TODO I don't think this provides a good way to handle targeted emotes when message_param doesn't exist
+	// Like, wouldn't you want to be able to scream at someone as a monkey?
+	if(params && message_param)
+		msg = select_param(user, params)
+		if(!msg)
+			to_chat(user, "<span class='warning'> '[params]' isn't a valid parameter for [key].")
+			return TRUE
+
+	msg = replace_pronoun(user, msg)
+
+	var/suppressed = FALSE
+
+	// Keep em quiet if they can't speak
+	if(!can_vocalize_emotes(user) && (emote_type & (EMOTE_MOUTH | EMOTE_AUDIBLE) || emote_type & (EMOTE_MOUTH | EMOTE_SOUND)))
+		var/noise_emitted = pick(muzzled_noises)
+		suppressed = TRUE
+		msg = "makes \a [noise_emitted] noise."
+
+	var/tmp_sound = get_sound(user)
+	var/sound_volume = get_volume(user)
+	// If our sound emote is forced by code, don't worry about cooldowns at all.
+	if(tmp_sound && should_play_sound(user, intentional) && sound_volume > 0)
+		play_sound_effect(user, intentional, tmp_sound, sound_volume)
+
+	if(msg)
+		if(isobserver(user))
+			log_ghostemote(msg, user)
+		else
+			log_emote(msg, user)
+
+		var/dchatmsg = "<b>[user]</b> [msg]"
+
+		var/user_turf = get_turf(user)
+		if(user.client)
+			for(var/mob/ghost as anything in GLOB.dead_mob_list)
+				if(!ghost.client || isnewplayer(ghost))
```

Comment:
I think this came directly from the /tg/ implementation. I'd imagine it might just be for safety, though I don't entirely know the nuances of new players.

Reply:
Yeah, I see the docs for `GLOB.dead_mob_list` note that it specifically excludes new players, so I'll just remove that check